qiskit/transpiler/passes/optimization/light_cone.py

```diff
@@ -0,0 +1,143 @@
+            same expectation value for the observable.
+        """
+        super().__init__()
+        self.observable = observable
+
+    @staticmethod
+    def _find_measurement_qubits(dag: DAGCircuit):
+        """For now, this method considers only final circuit measurements:
+        mid-circuit measurements are discarded.
+        """
+        qubits_measured = set()
+
+        # Iterate over the DAG nodes in topological order
+        for node in dag.topological_nodes():
+            # Check if the node is a measurement operation
+            if getattr(node, "name", False) == "measure":
+                qubits = set(node.qargs)
+
+                # Check if these qubits are used in any subsequent operations
+                is_final_measurement = True
+                for subsequent_node in dag.successors(node):
+                    if isinstance(subsequent_node, DAGOpNode):
+                        is_final_measurement = False
+                        break
+
+                qubits_measured |= qubits if is_final_measurement else set()
+        if not qubits_measured:
+            raise ValueError("No measurements found in the circuit.")
+        return qubits_measured
+
+    def _get_initial_lightcone(
+        self, dag: DAGCircuit
+    ) -> tuple[set[int], list[tuple[PauliGate, list[int]]]]:
+        """Returns the initial lightcone.
+        If obervable is `None`, the lightcone is the set of measured qubits.
+        If a Pauli observable is provided, the qubit corresponding to
+        the non-trivial Paulis define the lightcone.
+        """
+        if self.observable is None:
+            qubits_measured = self._find_measurement_qubits(dag)
+            light_cone = qubits_measured
+            light_cone_ops = [(PauliGate("Z"), [qubit_index]) for qubit_index in light_cone]
+        else:
+            # Check if the size of the observable matches the number of qubits in the circuit
+            if self.observable.num_qubits != dag.num_qubits():
+                raise ValueError(
+                    "Observable size does not match the number of qubits in the circuit."
+                )
+            # stripped_pauli = pauli_string.replace("I", "")
+            # if len(stripped_pauli_string) == 0:
+            #     raise ValueError("Observable is the identity operator.")
+            non_trivial_indices = [i for i, p in enumerate(self.observable) if p != Pauli("I")]
+            light_cone = [dag.qubits[i] for i in non_trivial_indices]
+            stripped_pauli_label = "".join(
+                [self.observable[i].to_label() for i in reversed(non_trivial_indices)]
+            )
+            light_cone_ops = [(PauliGate(stripped_pauli_label), light_cone)]
+
+        return set(light_cone), light_cone_ops
+
+    def run(self, dag: DAGCircuit):
+        """Run the LightCone pass on `dag`.
+
+        Args:
+            dag (DAGCircuit): the DAG to be optimized.
+
+        Returns:
+            DAGCircuit: the optimized DAG.
```

Comment:
You can skip the typehints here since if they are present in the signature 🙂 I'd also suggest to specify what "optimize" means here.
```suggestion
            dag: The DAG to reduce.

        Returns:
            The DAG reduced to the light cone of the observable.
```